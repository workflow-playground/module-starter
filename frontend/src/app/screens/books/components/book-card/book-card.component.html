<li class="col-span-1 flex flex-col divide-y divide-gray-200 rounded-lg bg-white text-center shadow min-h-[510px]">
  <div class="flex flex-1 flex-col p-8">
    <div class="mx-auto flex-shrink-0" *ngIf="thumbnail">
      <ds-image [height]="185" [src]="thumbnail"></ds-image>
    </div>
    <h3 class="mt-6 text-base font-bold text-gray-900">{{ title | bcEllipsis: 35 }}</h3>

    <dl class="mt-1 flex flex-grow flex-col justify-between">
      <dd class="text-sm text-gray-500">{{ description | bcEllipsis }}</dd>
      <dd class="mt-3" *ngIf="authors.length > 0">
        <span
          *ngFor="let author of authors"
          class="inline-flex items-center rounded-full bg-primary-p300-main m-1 px-2 py-1 text-xs font-medium text-primary-p100 ring-1 ring-inset ring-primary-p300"
          >{{ author }}</span
        >
      </dd>
    </dl>
  </div>
  <ng-content></ng-content>
</li>
