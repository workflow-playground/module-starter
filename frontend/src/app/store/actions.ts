import { actions as books } from './slices/books.store';
import { actions as collection } from './slices/collection.store';
import { actions as search } from './slices/search.store';
import { actions as auth } from './slices/auth.store';
import { actions as layout } from './slices/layout.store';
import { actions as loginPage } from './slices/login-page.store';
import { actions as user } from './slices/user.store';

export { books, collection, search, auth, loginPage, layout, user };
