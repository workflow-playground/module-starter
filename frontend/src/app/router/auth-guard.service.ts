import { inject } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';
import { map, take } from 'rxjs/operators';

import * as actions from '../store/actions';
import * as selectors from '../store/selectors';

export const authGuard = (): Observable<boolean> => {
  const store = inject(Store);

  return store.select(selectors.auth.selectLoggedIn).pipe(
    map(authed => {
      if (!authed) {
        store.dispatch(actions.auth.loginRedirect());
        return false;
      }

      return true;
    }),
    take(1),
  );
};
