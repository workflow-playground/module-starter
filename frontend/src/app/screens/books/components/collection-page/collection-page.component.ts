import { ChangeDetectionStrategy, Component } from '@angular/core';
import { Store } from '@ngrx/store';
import { IconColor, IconName } from '@ds24/elements';

import * as selectors from '../../../../store/selectors';

@Component({
  selector: 'ds-collection-page',
  changeDetection: ChangeDetectionStrategy.OnPush,
  templateUrl: './collection-page.component.html',
})
export class CollectionPageComponent {
  icons: typeof IconName = IconName;
  iconColors: typeof IconColor = IconColor;

  books$ = this.store.select(selectors.collection.selectBookCollection);

  constructor(private store: Store) {}
}
