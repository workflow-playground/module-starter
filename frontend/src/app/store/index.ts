import * as selectors from './selectors';
import * as actions from './actions';
import { State } from './state';

export {
  selectors,
  State,
  actions
}
