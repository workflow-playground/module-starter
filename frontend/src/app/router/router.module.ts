import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';

import { authGuard } from './auth-guard.service';

export const routes: Routes = [
  { path: '', redirectTo: '/books', pathMatch: 'full' },
  {
    path: 'books',
    loadChildren: () =>
      import('../screens/books/books.module').then((m) => m.BooksScreenModule),
    canActivate: [authGuard],
  },
  {
    path: 'login',
    loadChildren: () =>
      import('../screens/login/login.module').then((m) => m.LoginScreenModule)
  },
  {
    path: '**',
    loadChildren: () =>
      import('../screens/not-found/not-found.module').then((m) => m.NotFoundScreenModule)
  },
];

@NgModule({
  imports: [
    RouterModule.forRoot(routes, {
      useHash: true,
    }),
  ],
  exports: [RouterModule],
})
export class AppRouterModule {}
