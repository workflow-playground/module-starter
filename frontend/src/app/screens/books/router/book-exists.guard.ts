import { inject } from '@angular/core';
import { Router, ActivatedRouteSnapshot } from '@angular/router';
import { Store } from '@ngrx/store';
import { Observable, of } from 'rxjs';
import { catchError, map, switchMap, take, tap } from 'rxjs/operators';

import { GoogleBooksService } from '../../../services/google-books.service';
import * as selectors from '../../../store/selectors';
import * as actions from '../../../store/actions';

/**
 * Guards are hooks into the route resolution process, providing an opportunity
 * to inform the router's navigation process whether the route should continue
 * to activate this route. Guards must return an observable of true or false.
 */

export const bookExistsGuard = (route: ActivatedRouteSnapshot): Observable<boolean> => {
  const store = inject(Store);
  const googleBooks = inject(GoogleBooksService);
  const router = inject(Router);

  /**
   * This method checks if a book with the given ID is already registered
   * in the Store
   */
  function hasBookInStore(id: string): Observable<boolean> {
    return store.select(selectors.books.selectBookEntities).pipe(
      map(entities => !!entities[id]),
      take(1),
    );
  }

  /**
   * This method loads a book with the given ID from the API and caches
   * it in the store, returning `true` or `false` if it was found.
   */
  function hasBookInApi(id: string): Observable<boolean> {
    return googleBooks.retrieveBook(id).pipe(
      map(bookEntity => actions.books.loadBook({ book: bookEntity })),
      tap(action => store.dispatch(action)),
      map(book => !!book),
      catchError(() => {
        router.navigate(['/404']);
        return of(false);
      }),
    );
  }

  /**
   * `hasBook` composes `hasBookInStore` and `hasBookInApi`. It first checks
   * if the book is in store, and if not it then checks if it is in the
   * API.
   */
  function hasBook(id: string): Observable<boolean> {
    return hasBookInStore(id).pipe(
      switchMap(inStore => {
        if (inStore) {
          return of(inStore);
        }

        return hasBookInApi(id);
      }),
    );
  }

  /**
   * This is the actual method the router will call when our guard is run.
   *
   * Our guard waits for the collection to load, then it checks if we need
   * to request a book from the API or if we already have it in our cache.
   * If it finds it in the cache or in the API, it returns an Observable
   * of `true` and the route is rendered successfully.
   *
   * If it was unable to find it in our cache or in the API, this guard
   * will return an Observable of `false`, causing the router to move
   * on to the next candidate route. In this case, it will move on
   * to the 404 page.
   */
  return hasBook(route.params['id']);
};
