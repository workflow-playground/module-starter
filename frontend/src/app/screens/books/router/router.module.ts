import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import { CollectionPageComponent } from '../components/collection-page.component';
import { bookExistsGuard } from './book-exists.guard';
import { BookDetailsPageComponent } from '../components/book-details-page.component';
import { FindBookPageComponent } from '../components/find-book-page.component';

export const routes: Routes = [
  {
    path: 'find',
    component: FindBookPageComponent,
    data: { title: 'Find book', page: 'books' },
  },
  {
    path: ':id',
    component: BookDetailsPageComponent,
    canActivate: [bookExistsGuard],
    data: { title: 'Book details', page: 'collection' },
  },
  {
    path: '',
    component: CollectionPageComponent,
    data: { title: 'Collection', page: 'collection' },
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
})
export class BooksRoutingModule {}
