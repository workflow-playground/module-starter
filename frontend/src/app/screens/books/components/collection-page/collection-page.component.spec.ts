import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { IconModule, ImageModule } from '@ds24/elements';
import { TranslateModule } from '@ngx-translate/core';
import { RouterTestingModule } from '@angular/router/testing';

import { CollectionPageComponent } from './collection-page.component';
import { BookCardComponent } from '../book-card/book-card.component';
import { State, selectors } from '../../../../store';
import { bookMock } from '../../../../types/book.types.mock';
import { PipesModule } from '../../../../shared/pipes/pipes.module';

describe('Collection Page', () => {
  let fixture: ComponentFixture<CollectionPageComponent>;
  let mockStore: MockStore<State>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [IconModule, PipesModule, ImageModule, TranslateModule.forRoot(), RouterTestingModule],
      declarations: [CollectionPageComponent, BookCardComponent],
      providers: [provideMockStore()],
    });

    fixture = TestBed.createComponent(CollectionPageComponent);
    mockStore = TestBed.inject(MockStore);
    mockStore.overrideSelector(selectors.collection.selectBookCollection, [bookMock]);

    jest.spyOn(mockStore, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
