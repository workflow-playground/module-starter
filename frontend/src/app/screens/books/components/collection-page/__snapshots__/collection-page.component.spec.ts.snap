// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Page should compile 1`] = `
<ds-collection-page
  books$={[Function _Store]}
  iconColors={[Function Object]}
  icons={[Function Object]}
  store={[Function _MockStore]}
>
  <ul
    class="grid grid-cols-1 gap-6 sm:grid-cols-2 md:grid-cols-3"
    role="list"
  >
    <ds-book-card>
      <li
        class="col-span-1 flex flex-col divide-y divide-gray-200 rounded-lg bg-white text-center shadow min-h-[510px]"
      >
        <div
          class="flex flex-1 flex-col p-8"
        >
          <div
            class="mx-auto flex-shrink-0"
          >
            <ds-image>
              <div
                class="image-container"
                style="height: 185px; width: 100%;"
              >
                <svg
                  height="40px"
                  id="loader-1"
                  style="enable-background: new 0 0 50 50;"
                  version="1.1"
                  viewBox="0 0 50 50"
                  width="40px"
                  x="0px"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                  xmlns:xlink="http://www.w3.org/1999/xlink"
                  y="0px"
                >
                  <path
                    background="deeppink"
                    d="M43.935,25.145c0-10.318-8.364-18.683-18.683-18.683c-10.318,0-18.683,8.365-18.683,18.683h4.068c0-8.071,6.543-14.615,14.615-14.615c8.072,0,14.615,6.543,14.615,14.615H43.935z"
                    fill="#31373c"
                  >
                    <animatetransform
                      attributeName="transform"
                      attributeType="xml"
                      dur="0.6s"
                      from="0 25 25"
                      repeatCount="indefinite"
                      to="360 25 25"
                      type="rotate"
                    />
                  </path>
                </svg>
                <img
                  alt="ds-image"
                  class="hidden"
                  src="string"
                  style="height: 185px; width: 100%;"
                />
              </div>
            </ds-image>
          </div>
          <h3
            class="mt-6 text-base font-bold text-gray-900"
          >
            title
          </h3>
          <dl
            class="mt-1 flex flex-grow flex-col justify-between"
          >
            <dd
              class="text-sm text-gray-500"
            >
              description
            </dd>
            <dd
              class="mt-3"
            >
              <span
                class="inline-flex items-center rounded-full bg-primary-p300-main m-1 px-2 py-1 text-xs font-medium text-primary-p100 ring-1 ring-inset ring-primary-p300"
              >
                author
              </span>
            </dd>
          </dl>
        </div>
        <div
          class="-mt-px flex divide-x divide-gray-200"
        >
          <div
            class="flex w-0 flex-1"
          >
            <a
              class="relative -mr-px inline-flex w-0 flex-1 items-center justify-center gap-x-3 rounded-bl-lg border border-transparent py-4 text-sm font-bold text-neutral-n500 hover:text-primary-p500 overflow-hidden"
              href="/books/1"
            >
              <ds-icon>
                <file-icon>
                  <svg
                    fill="none"
                    height="20px"
                    style="height: 20px; width: 20px;"
                    viewBox="0 0 16 16"
                    width="20px"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      opacity="0.6"
                    >
                      <path
                        class="fill-primary-p400"
                        d="M3.3335 1.33325V14.6666H13.3335V4.71274L13.1935 4.55377L10.1935 1.47684L10.0385 1.33325H3.3335ZM4.3335 2.35889H9.3335V5.43582H12.3335V13.6409H4.3335V2.35889ZM10.3335 3.07684L11.6135 4.41018H10.3335V3.07684Z"
                      />
                    </g>
                  </svg>
                </file-icon>
              </ds-icon>
               BOOKS.COLLECTION.SHOW_DETAILS 
            </a>
          </div>
        </div>
      </li>
    </ds-book-card>
  </ul>
</ds-collection-page>
`;
